### PR TITLE
STORM-4075 Supprt mTLS between Storm and ZK

### PR DESCRIPTION
**Pull request by purushah**
## What is the purpose of the change

*(Explain why we should have this change)*

## How was the change tested

*(Explain what tests did you do to verify the code change)*

### CODE REVIEW

File: storm-client/src/jvm/org/apache/storm/Config.java [hunk 1]
```diff
@@ -1419,6 +1419,25 @@ public class Config extends HashMap<String, Object> {
      */
     @IsString
     public static final String STORM_ZOOKEEPER_TOPOLOGY_AUTH_SCHEME = "storm.zookeeper.topology.auth.scheme";
+
+    /** Enable SSL/TLS for ZooKeeper client connection. */
+    @IsBoolean
+    public static final String ZK_SSL_ENABLE = "storm.zookeeper.ssl.enable";
```
Comment (rzo1) [inline storm-client/src/jvm/org/apache/storm/Config.java:1425]: Can we rename those constants starting with `ZK` to be consistent with the rest of the constant names? They all are something like `STORM_ZOOKEEPER_XY`?

File: storm-client/src/jvm/org/apache/storm/utils/CuratorUtils.java [hunk 2]
```diff
@@ -84,6 +89,120 @@ protected static void setupBuilder(CuratorFrameworkFactory.Builder builder, fina
...
+     * Configure ZooKeeper Client with SSL/TLS connection.
+     * @param zkClientConfig ZooKeeper Client configuration
+     * @param truststoreKeystore The truststore and keystore configs
+     */
+    private static void setSslConfiguration(ZKClientConfig zkClientConfig,
+                                            TruststoreKeystore truststoreKeystore) throws ConfigurationException {
+        setSslConfiguration(zkClientConfig, new ClientX509Util(), truststoreKeystore);
+    }
+
+    private static void setSslConfiguration(ZKClientConfig zkClientConfig,
```
Comment (rzo1) [inline storm-client/src/jvm/org/apache/storm/utils/CuratorUtils.java:115]: What is the reason for having two methods here? Both are `private`, so supplying a custom `ClientX509Util` isn't possible anyway?

File: storm-client/src/jvm/org/apache/storm/utils/CuratorUtils.java [hunk 3]
```diff
@@ -84,6 +89,120 @@ protected static void setupBuilder(CuratorFrameworkFactory.Builder builder, fina
...
+                    "The truststore password parameter is empty for the ZooKeeper client connection" + ".");
+        }
+    }
+
+
+    /**
+     * Helper class to contain the Truststore/Keystore paths for the ZK client connection over
+     * SSL/TLS.
+     */
+    public static class TruststoreKeystore {
```
[the remark this answers is not in the document]
Comment (rzo1) [reply storm-client/src/jvm/org/apache/storm/utils/CuratorUtils.java:167]: might be `final` as well.

File: storm-client/test/jvm/org/apache/storm/utils/CuratorUtilsTest.java [hunk 4]
```diff
@@ -18,22 +18,46 @@
...
+import org.apache.curator.test.TestingServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 
 import static org.junit.jupiter.api.Assertions.assertArrayEquals;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 public class CuratorUtilsTest {
+    public static final Logger LOG = LoggerFactory.getLogger(CuratorUtilsTest.class);
```
Comment (rzo1) [inline storm-client/test/jvm/org/apache/storm/utils/CuratorUtilsTest.java:48]: Can all be `private` ?

**Comment by purushah** [comment]
Test this locally with ZK running with mTls.
Zk setting

```
client.certReload=true
secureClientPort=50513
serverCnxnFactory=org.apache.zookeeper.server.NettyServerCnxnFactory

clientCnxnSocket=org.apache.zookeeper.ClientCnxnSocketNetty
ssl.keyStore.location=/tmp/ssl/server.keystore.jks
ssl.keyStore.password=password
ssl.trustStore.location=/tmp/ssl/server.truststore.jks
ssl.trustStore.password=password
ssl.hostnameVerification=false
ssl.clientAuth=need

```

Storm setting

```
storm.zookeeper.ssl.enable: true
storm.zookeeper.ssl.keystore.path: /tmp/ssl/client.keystore.jks
storm.zookeeper.ssl.keystore.password: password
storm.zookeeper.ssl.truststore.path:/tmp/ssl/client.truststore.jks
storm.zookeeper.ssl.truststore.password: password
storm.zookeeper.ssl.hostnameVerification: false

storm.zookeeper.servers:
  - "127.0.0.1"
storm.zookeeper.port: 50513

```

**Review by rzo1 — APPROVED** [review]

**Review by snikhil5 — COMMENTED** [review]
lgtm. We need to update the documents too https://storm.apache.org/releases/current/SECURITY

**Comment by purushah** [comment]
Thanks, @snikhil5 will update the Doc as part of https://issues.apache.org/jira/browse/STORM-4079

**Pull request merged**